testing/src/main/java/uk/gov/service/payments/commons/testing/port/PortFactory.java

```diff
@@ -17,4 +17,18 @@ public static int findFreePort() {
         }
         return port;
     }
+
+    public static int findFreePortWithoutCollision(int... existingPorts) {
+        boolean collision = true;
+        int port = 0;
+
+        while (collision) {
+            port = findFreePort();
+            collision = false;
+            for (int existingPort : existingPorts) {
+                collision = collision || port == existingPort;
+            }
+        }
+        return port;
+    }
```

Comment:
You could also do something like this:

```java
public static int findFreePortWithoutCollision(int... existingPorts) {
    int[] sortedExistingPorts = existingPorts.clone();
    Arrays.sort(sortedExistingPorts);

    return Stream.generate(PortFactory::findFreePort)
            .filter(port -> Arrays.binarySearch(sortedExistingPorts, port) < 0)
            .findFirst()
            .orElseThrow(() -> new IllegalStateException("This can never happen"));
}

```

Reply:
Binary search is much preferred as it has a time complexity of O(logn) but iterating through an array is potentially O(n).

Reply:
Sorting is still O(nlogn) though, so I think whichever you do you can't get better than that. However the way I did it could be O(n^2) in the worst case as the probability of a collision is linear in the number of existing ports. In practice I think we're talking about such a small value of n that the wallclock time will always be negligible though (although I know it's dangerous to make an assumption about that)

Reply:
> we're talking about such a small value of n

True! I'm not gonna be unhappy if you merge this PR 😄

Reply:
Equally I'm more than happy to change it. To be perfectly honest it might be better to have a class that would by default track ports returned from `findFreePort` and always avoid collisions. While I like it being a pure function, that feels like the Java way to do it

Reply:
It's up to you. But yeah, if a port is always to a Set from `findFreePort` we wouldn't have to sort in `findFreePortWithoutCollision`.